### PR TITLE
Query Creator Attribute Groups

### PR DESCRIPTION
**Pull request by kespinola**
### Goal
A client can query the set of attribute groups associated to a creator's nfts with count of each variant.
A client can filter the nfts root query by attributes.

```graphql
creator(address: "{address}") {
  attributeGroups {
    name
     variants {
         name
         count
     }
   }
}
```

### CODE REVIEW

File: crates/core/src/db/models.rs [hunk 1]
```diff
@@ -263,8 +263,22 @@ pub struct File<'a> {
...
+pub struct MetadataAttributeWrite<'a> {
+    /// Metadata address
+    pub metadata_address: Cow<'a, str>,
+    /// Attribute name
+    pub name: Option<Cow<'a, str>>,
+    /// Attribute value
+    pub value: Option<Cow<'a, str>>,
+    /// Attribute trait type
+    pub trait_type: Option<Cow<'a, str>>,
+}
```
Comment (kespinola) [inline crates/core/src/db/models.rs:277]: When using generated fields within the db there needs to be a separate write vs read model.

Comment (kespinola) [reply crates/core/src/db/models.rs:277]: The write has insertable and query the Queryable derive.

In my reading on this the docs also suggest not mixing a struct as both queryable and insertable. Thoughts @ray-kast

File: crates/core/src/db/models.rs [hunk 2]
```diff
@@ -273,6 +287,8 @@ pub struct Attribute<'a> {
     pub value: Option<Cow<'a, str>>,
     /// Attribute trait type
     pub trait_type: Option<Cow<'a, str>>,
+    /// Attribute generated id
+    pub id: Cow<'a, uuid::Uuid>,
```
Comment (kespinola) [inline crates/core/src/db/models.rs:291]: The model types can not be deisel types. Must always be rust type hence pulling in the uuid crate.

File: crates/graphql/src/schema.rs [hunk 3]
```diff
@@ -5,13 +5,92 @@ use dataloader::{non_cached::Loader, BatchFn};
...
+                });
+            }
+
+            attribute_groups.push(Property {
+                name: attribute_group_name,
+                variants,
+            });
+        }
+
+        attribute_groups
```
Comment (kespinola) [inline crates/graphql/src/schema.rs:78]: My ideal for this is to have database views that recalculated based on computed columns but this should get us solid traction. Can cache the queries with a ttl as the next improvement.

Comment (kespinola) [reply crates/graphql/src/schema.rs:78]: I opted for commuting with rust vs making a bunch of select where count queries within sql.

File: crates/indexer/src/bits/metadata_uri.rs [hunk 4]
```diff
@@ -119,11 +119,15 @@ enum ValueDataType {
...
-impl fmt::Display for ValueDataType {
+impl Display for ValueDataType {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{:?}", self)
+        match self {
+            Self::Number(n) => Display::fmt(n, f),
+            Self::String(s) => Display::fmt(s, f),
+        }
     }
 }
```
Comment (kespinola) [inline crates/indexer/src/bits/metadata_uri.rs:129]: Patch @ray-kast put together to fix the writing of the values to the db.

File: crates/graphql/src/schema.rs [hunk 5]
```diff
@@ -5,13 +5,92 @@ use dataloader::{non_cached::Loader, BatchFn};
...
+    fn address(&self) -> String {
+        self.address.clone()
+    }
+
+    pub fn attribute_groups(&self, context: &AppContext) -> Vec<Property> {
+        let conn = context.db_pool.get().unwrap();
+
+        let metadatas: Vec<String> = metadata_creators::table
+            .select(metadata_creators::metadata_address)
+            .filter(metadata_creators::creator_address.eq(self.address.clone()))
```
Comment (kespinola) [inline crates/graphql/src/schema.rs:31]: This is a nifty feature of graphql is using the context of the parent node in downstream lookups. very powerful.

File: crates/core/src/db/models.rs (hunk 1 above)
Comment (ray-kast) [reply crates/core/src/db/models.rs:277]: Yeah, for structs where we have write-default fields separate would be good.  Diesel generally recommend making purpose-built models, but I merged them where possible to reduce boilerplate.

File: crates/graphql/src/schema.rs (hunk 3 above)
Comment (ray-kast) [reply crates/graphql/src/schema.rs:78]: Yeah, diesel only really supports collating rows outside of SQL.

Comment (ray-kast) [reply crates/graphql/src/schema.rs:78]: @kespinola can i get a rundown on what this code is doing?

**Review by ray-kast — COMMENTED** [review]
Looks good overall, although I'm curious if the double `for`-loop can be eliminated.

**Comment by kespinola** [comment]
@ray-kast its taking all the attributes for the metadata associated to the creator and reducing them to a map that looks like this

```
{"Size": {"Small (S)": 14, "Extra Large (XL)": 4, "Medium (M)": 9}, "Type": {"Land": 44}, "Light Source": {"Yes": 28}, "Edition": {"Founders Edition (FE)": 11, "Founders Edition": 5}, "Dynamic Bones": {"Yes": 28}, "Height": {"50m ": 1, "25n": 1, "25m": 33, "50m": 8, "25": 1}, "Lip Sync": {"Yes": 28}, "MBs": {"100": 7, "200": 4, "75": 7, "600": 4, "250": 5}, "Custom Shaders": {"Yes": 28}, "Full Body Tracking": {"Yes": 28}, "Size m²": {"600": 35, "1500": 9}, "Finger Tracking": {"Yes": 28}, "Parcel Size": {"Medium (M)": 35, "Extra Large (XL)": 9}}
```

Then transforming into a list structure for the ui which is more usable but less efficient for doing the counting.

File: crates/graphql/src/schema.rs (hunk 3 above)
Comment (kespinola) [reply crates/graphql/src/schema.rs:78]: @ray-kast its taking all the attributes for the metadata associated to the creator and reducing them to a map that looks like this
```
{"Size": {"Small (S)": 14, "Extra Large (XL)": 4, "Medium (M)": 9}, "Type": {"Land": 44}, "Light Source": {"Yes": 28}, "Edition": {"Founders Edition (FE)": 11, "Founders Edition": 5}, "Dynamic Bones": {"Yes": 28}, "Height": {"50m ": 1, "25n": 1, "25m": 33, "50m": 8, "25": 1}, "Lip Sync": {"Yes": 28}, "MBs": {"100": 7, "200": 4, "75": 7, "600": 4, "250": 5}, "Custom Shaders": {"Yes": 28}, "Full Body Tracking": {"Yes": 28}, "Size m²": {"600": 35, "1500": 9}, "Finger Tracking": {"Yes": 28}, "Parcel Size": {"Medium (M)": 35, "Extra Large (XL)": 9}}
```
Then transforming into a list structure for the ui which is more usable but less efficient for doing the counting.

Comment (ray-kast) [inline crates/graphql/src/schema.rs:78]: Doing something like the following would be more idiomatic Rust:
```
    let properties = metadata_attributes
        .into_iter()
        .fold(
            HashMap::new(),
            |mut groups,
             MetadataAttribute {
                 trait_type, value, ..
             }| {
                *groups
                    .entry(trait_type)
                    .or_insert_with(HashMap::new)
                    .entry(value)
                    .or_insert(0) += 1;

                groups
            },
        )
        .into_iter()
        .map(|(name, vars)| Property {
            name,
            variants: vars
                .into_iter()
                .map(|(name, count)| PropertyVariant { name, count })
                .collect(),
        })
        .collect::<Vec<_>>();
```

**Review by ray-kast — CHANGES_REQUESTED** [review]

**Review by ray-kast — APPROVED** [review]
:+1: :smile:

**Pull request merged**